int32 coming_from
float32 x
float32 y
float32 z